### PR TITLE
New Order Types FOK/IOC #622

### PR DESCRIPTION
**Pull request by rsercano**
Hello guys,

this is an initial commit for #622, so it's not finished yet. Even though I don't feel like I'm lost yet, but feel like gotta lose soon, so please let me know what you think.

I've changed `buy` and `sell` commands a bit, now I guess they look more meaningful, instead of controlling `price` arg to decide now we accept another arg for it, so it can be `lmt`, `ioc`, `fok`, `mkt` which's clearer I guess. 

It would be great if you can give me some directions from here, just maybe shorten to understand further, thanks in advance !

p.s. I'm aware tests are failing, just wanted some directions from you guys.

### CODE REVIEW

**Comment by sangaman** [comment]
For the command line, I feel like the `fok`/`ioc` orders would probably not be as common. I'd be hesitant to change the regular way to add limit orders. Maybe add `--fok` and `--ioc` flags that can be set manually? I'm actually thinking it might make sense to have `[stream]` be a flag like `--stream` in the same way as well.

Fill-or-kill will be complicated to implement properly I think, but you can definitely go for it. If you want to just do `ioc` first that might be easier.

**Comment by sangaman** [comment]
One thing that we'd need to change for `fok` is allowing multiple swap deals with the same `rHash`. Right now it's basically the key by which we identify swap deals. It'd require db changes as well as changes to several places in the code. We could maybe make swaps unique according to the pair of `rHash` and the maker `orderId`.

**Comment by kilrau** [comment]
> For the command line, I feel like the `fok`/`ioc` orders would probably not be as common. I'd be hesitant to change the regular way to add limit orders. Maybe add `--fok` and `--ioc` flags that can be set manually?

Yep, separate flags with the regular limit order commands NOT changing (a bit messy on gitter, check chats from Jan 04)

**Comment by offerm** [comment]
Why we need multiple swap deals having the same hash?

**Comment by sangaman** [comment]
> Why we need multiple swap deals having the same hash?

@offerm I think I was mistaken, after thinking about this some more it shouldn't be necessary. But I figure we would need to make sure that we have HTLC's to us for all portions of the `fok` order before releasing the preimages, which is still a significant change.

**Comment by sangaman** [comment]
@kilrau I figure we would need to find all matches and get all swaps accepted and up to the point where the HTLCs are set up but right before we release the preimage. We can't execute any trades before we get to that point, and if we can't get there we have to discard the trade entirely.

It would also be tricky because we'd need to hold off on executing the internal match trades immediately as we always do now. I'm actually not sure how to go about this part, because if we find 2 matches, one with an internal order and one with an external order, we can't execute the internal order until the swap for the external order is accepted and set up. But since setting up the swap is not instant, in the meantime it's possible that the internal order gets canceled or filled by a different order.

In other words I think fill-or-kill orders introduce a great deal of complexity and probably shouldn't be a priority.

**Comment by kilrau** [comment]
Agree, let's move [this](https://github.com/ExchangeUnion/xud/issues/622) to `alpha.11`

**Comment by kilrau** [comment]
According to @rsercano ([comment](https://github.com/ExchangeUnion/xud/issues/622#issuecomment-457147216)) close this PR?
